### ISSUE TITLE
Issue with Set-TimeZoneFromIP

### ISSUE THREAD

**Issue by JHBDO** [OSDeploy/OSD#110]
**Describe the bug**
It appears that the $TimeZone variable in the Set-TimeZoneFromIP function is an array that contains two items.  The command Set-TimeZone -Name $TimeZone throws an error.

**To Reproduce**
1. In OOBE, call the Set-TimeZoneFromIP function.
2. See error in Additional context below

**Expected behavior**
Function should set the timezone without an error.

**Desktop (please complete the following information):**
 - OS: Windows 11 23H2
 - OSD version 24.2.4.1

**Additional context**
When I call Set-TimeZoneFromIP from my  script in the OOBE phase, it returns the error:
Set-TimeZone : Cannot convert 'System.Object[]' to the type 'System.String' required by the parameter 'Name'.  Specified method is not supported.

I'm in the Eastern timezone, and when I debug the function, I see that the $TimeZone variable is an array containing:
Eastern Standard Time
Eastern Daylight Time

Based on the error message, I'm assuming this should be returning only 1 item as a string, and not an array.  $TimeZone.GetType() says the variable is an object with BaseType System.Array.

**Comment by JHBDO** [comment]
Thanks Gary.  As a temporary workaround, I added this code to Set-TimeZoneFromIP.  It compares the time zone offset value in order to narrow it down to one.  It was generated by Microsoft Copilot, so there's probably a more efficient way to do it.  :-)

##Additionally matches on time zone offset value to fix an issue with mutiple Daylight/Standard values being returned
	$offset = ($TimeZoneAPIInfo.utc_offset).Substring(0,3)
        if ($offset.Substring(1,1) -eq "0") {
            $offset = $offset.Substring(0,1) + $offset.Substring(2,1)
        }
	return ($TimeZoneData | Where-Object {$_.utc -match $TimeZoneAPIInfo.timezone -and $_.offset -match $offset}).value

**Comment by gwblok** [comment]
Thanks, can you try the updated code here: https://github.com/OSDeploy/OSD/blob/master/Public/OSDCloudTS/Get-TimeZoneFromIP.ps1

The problem was, while your updated code worked in your timezone, it didn't work in my timezone, it returned zero results.  
So I used your code and added another if statement.  Hopefully that works for most people.

I'm tempted to just have it so you can set a variable with the time zone, and it will apply that time zone for you, instead of trying to do a lookup based on the IP address. (when doing zero touch deployments)

Basically make a table based on this data: https://learn.microsoft.com/en-us/windows-hardware/manufacture/desktop/default-time-zones?view=windows-11.

$Global:MyOSDCloud.TimZone = "Eastern Standard Time"

Thoughts on that?

**Comment by JHBDO** [comment]
For me personally, having the time zone set based on IP geo is a must, as we have locations across the globe.  Using the IP really cuts down on the number of configs I need for ZTI.

________________________________

**Comment by gwblok** [comment]
with the updated changes, has this been working for you now?
I don't have a way to test most Time Zones, so hopefully as other have issues, we can just fix those bugs as they come

**Comment by gwblok** [comment]
Considering Closed until someone else has an issue in their TimeZone.

**Comment by Abe-Telo** [comment]
Testing this in Windows. we get the following, Did not test in WinPE. IS this expected to happen?
```
PS C:\Users\Administrator> Get-TimeZoneFromIP
PS C:\Users\Administrator> Set-TimeZoneFromIP
Setting Timezone to Eastern Standard Time Eastern Daylight Time - Online Mode in Full OS
Set-TimeZone : Cannot convert 'System.Object[]' to the type 'System.String' required by parameter 'Name'. Specified
method is not supported.
At C:\Program Files\WindowsPowerShell\Modules\OSD\25.1.16.2\Public\OSDCloudTS\Set-TimeZoneFromIP.ps1:1499 char:32
+             Set-TimeZone -Name $TimeZone
+                                ~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Set-TimeZone], ParameterBindingException
    + FullyQualifiedErrorId : CannotConvertArgument,Microsoft.PowerShell.Commands.SetTimeZoneCommand
```

**Comment by gwblok** [comment]
@Abe-Telo, not sure what to tell ya, works for me

PS C:\Users\GaryBlok> Get-TimeZoneFromIP
Central Standard Time
PS C:\Users\GaryBlok> Set-TimeZoneFromIP
Setting Timezone to Central Standard Time - Online Mode in Full OS

What version of PowerShell are you using?
